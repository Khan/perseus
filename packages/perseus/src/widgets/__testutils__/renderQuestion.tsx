import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
