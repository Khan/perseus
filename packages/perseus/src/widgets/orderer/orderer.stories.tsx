import {generateTestPerseusItem} from "@khanacademy/perseus-core";

import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";

import {question1, question2, questionWithImages} from "./orderer.testdata";

import type {Meta, StoryObj} from "@storybook/react";

const meta: Meta = {
    title: "Perseus/Widgets/Orderer",
    component: ServerItemRendererWithDebugUI,
};
export default meta;

type Story = StoryObj<typeof ServerItemRendererWithDebugUI>;

export const Question1: Story = {
    args: {
        item: generateTestPerseusItem({question: question1}),
    },
};

export const QuestionWithImages: Story = {
    args: {
        item: generateTestPerseusItem({question: questionWithImages}),
    },
};

export const Answerless: Story = {
    args: {
        item: generateTestPerseusItem({question: question2}),
        startAnswerless: true,
    },
};
