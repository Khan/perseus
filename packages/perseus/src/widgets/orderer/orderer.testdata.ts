import type {PerseusRenderer} from "../../perseus-types";

export const question1: PerseusRenderer = {
    content:
        "**Without using a calculator, put the numbers in order  from least to greatest.**  \n\n[[\u2603 orderer 1]]",
    images: {},
    widgets: {
        "orderer 1": {
            version: {major: 0, minor: 0},
            type: "orderer",
            graded: true,
            options: {
                otherOptions: [],
                layout: "horizontal",
                options: [
                    {content: "$10.9$", images: {}, widgets: {}},
                    {content: "$11$", images: {}, widgets: {}},
                    {content: "$\\sqrt{120}$", images: {}, widgets: {}},
                ],
                correctOptions: [
                    {content: "$10.9$", images: {}, widgets: {}},
                    {content: "$\\sqrt{120}$", images: {}, widgets: {}},
                    {content: "$11$", images: {}, widgets: {}},
                ],
                height: "normal",
            },
        },
    },
};

export const question2: PerseusRenderer = {
    content:
        "**Without using a calculator, put the numbers in order  from least to greatest.**  \n\n[[\u2603 orderer 1]]",
    images: {},
    widgets: {
        "orderer 1": {
            version: {major: 0, minor: 0},
            type: "orderer",
            graded: true,
            options: {
                otherOptions: [],
                layout: "horizontal",
                options: [
                    {content: "1", images: {}, widgets: {}},
                    {content: "3", images: {}, widgets: {}},
                    {content: "2", images: {}, widgets: {}},
                ],
                correctOptions: [
                    {content: "1", images: {}, widgets: {}},
                    {content: "2", images: {}, widgets: {}},
                    {content: "3", images: {}, widgets: {}},
                ],
                height: "normal",
            },
        },
    },
};

export const questionWithImages: PerseusRenderer = {
    content: "**Put $6$ flowers in the box.**\n\n[[â˜ƒ orderer 1]]",
    images: {},
    widgets: {
        "orderer 1": {
            graded: true,
            options: {
                correctOptions: [
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                ],
                height: "auto",
                layout: "horizontal",
                options: [
                    {
                        content:
                            "![](https://ka-perseus-graphie.s3.amazonaws.com/b7a6f30d245d186cf42961677ddafa118fef5fdd.png)",
                        widgets: {},
                        images: {},
                    },
                ],
                otherOptions: [],
            },
            type: "orderer",
            version: {
                major: 0,
                minor: 0,
            },
        },
    },
};
