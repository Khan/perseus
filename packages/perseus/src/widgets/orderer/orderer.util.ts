import type {
    PerseusOrdererWidgetOptions,
    PerseusRenderer,
} from "@khanacademy/perseus-core";

export type OrdererPublicWidgetOptions = {
    // All of the options available to the user. Place the cards in the correct order. The same card can be used more than once in the answer but will only be displayed once at the top of a stack of identical cards.
    options: ReadonlyArray<PerseusRenderer>;
    // "normal" for text options.  "auto" for image options.
    height: "normal" | "auto";
    // Use the "horizontal" layout for short text and small images. The "vertical" layout is best for longer text (e.g. proofs).
    layout: "horizontal" | "vertical";
};

function getOrdererPublicOptions(
    options: PerseusOrdererWidgetOptions,
): OrdererPublicWidgetOptions {
    return {
        options: options.options,
        height: options.height,
        layout: options.layout,
    };
}

export default getOrdererPublicOptions;
