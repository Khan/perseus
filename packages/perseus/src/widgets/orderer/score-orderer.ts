import _ from "underscore";

import validateOrderer from "./validate-orderer";

import type {PerseusScore} from "../../types";
import type {
    PerseusOrdererScoringData,
    PerseusOrdererUserInput,
} from "../../validation.types";

export function scoreOrderer(
    userInput: PerseusOrdererUserInput,
    scoringData: PerseusOrdererScoringData,
): PerseusScore {
    const validateError = validateOrderer(userInput);
    if (validateError) {
        return validateError;
    }

    const correct = _.isEqual(
        userInput.current,
        scoringData.correctOptions.map((option) => option.content),
    );

    return {
        type: "points",
        earned: correct ? 1 : 0,
        total: 1,
        message: null,
    };
}
