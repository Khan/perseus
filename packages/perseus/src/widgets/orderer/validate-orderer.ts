import type {PerseusScore} from "../../types";
import type {PerseusOrdererUserInput} from "../../validation.types";

/**
 * Checks user input from the orderer widget to see if the user has started
 * ordering the options, making the widget scorable.
 * @param userInput
 * @see `scoreOrderer` for more details.
 */
function validateOrderer(
    userInput: PerseusOrdererUserInput,
): Extract<PerseusScore, {type: "invalid"}> | null {
    if (userInput.current.length === 0) {
        return {
            type: "invalid",
            message: null,
        };
    }

    return null;
}

export default validateOrderer;
