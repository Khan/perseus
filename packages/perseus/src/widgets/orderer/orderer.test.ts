import {scorePerseusItem} from "@khanacademy/perseus-score";
import {act, screen} from "@testing-library/react";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {getAnswerfulItem, getAnswerlessItem} from "../../util/test-utils";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {question2} from "./orderer.testdata";

import type {APIOptions} from "../../types";
import type {PerseusOrdererWidgetOptions} from "@khanacademy/perseus-core";

const ordererOptions: PerseusOrdererWidgetOptions = {
    otherOptions: [],
    layout: "horizontal",
    options: [
        {content: "1", images: {}, widgets: {}},
        {content: "3", images: {}, widgets: {}},
        {content: "2", images: {}, widgets: {}},
    ],
    correctOptions: [
        {content: "1", images: {}, widgets: {}},
        {content: "2", images: {}, widgets: {}},
        {content: "3", images: {}, widgets: {}},
    ],
    height: "normal",
};

describe("orderer widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };

        // Act
        const {container} = renderQuestion(question2, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: true,
        };

        // Act
        const {container} = renderQuestion(question2, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    test("the answerless test data doesn't contain answers", () => {
        // Arrange / Act / Assert
        expect(
            getAnswerlessItem("orderer", ordererOptions).question.widgets[
                "orderer 1"
            ].options.correctOptions,
        ).toBeUndefined();
        expect(
            getAnswerlessItem("orderer", ordererOptions).question.widgets[
                "orderer 1"
            ].options.otherOptions,
        ).toBeUndefined();
    });

    describe.each([
        ["answerless", getAnswerlessItem("orderer", ordererOptions)],
        ["answerful", getAnswerfulItem("orderer", ordererOptions)],
    ])("given %s options", (_, {question}) => {
        it("renders", async () => {
            // Arrange / Act
            renderQuestion(question);

            // Assert
            expect(await screen.findByText("1")).toBeInTheDocument();
            expect(await screen.findByText("2")).toBeInTheDocument();
        });

        it("can be answered", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            const [orderer] = renderer.findWidgets("orderer 1");
            act(() => orderer.setListValues(["37"]));
            const userInput = renderer.getUserInputMap();

            // Assert
            expect(userInput).toEqual({"orderer 1": {current: ["37"]}});
        });

        it("can give an invalid score", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            const userInput = renderer.getUserInputMap();
            const score = scorePerseusItem(
                getAnswerfulItem("orderer", ordererOptions).question,
                userInput,
                "en",
            );

            // Assert
            expect(score).toHaveInvalidInput();
        });

        it("can be answered correctly", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            const [orderer] = renderer.findWidgets("orderer 1");

            act(() => orderer.setListValues(["1", "2", "3"]));
            const userInput = renderer.getUserInputMap();

            const score = scorePerseusItem(
                getAnswerfulItem("orderer", ordererOptions).question,
                userInput,
                "en",
            );

            // Assert
            expect(score).toHaveBeenAnsweredCorrectly();
        });

        it("can be scored incorrectly", () => {
            // Arrange
            const {renderer} = renderQuestion(question);

            // Act
            const [orderer] = renderer.findWidgets("orderer 1");

            act(() => orderer.setListValues(["2", "3", "1"]));
            const userInput = renderer.getUserInputMap();

            const score = scorePerseusItem(
                getAnswerfulItem("orderer", ordererOptions).question,
                userInput,
                "en",
            );

            // Assert
            expect(score).toHaveBeenAnsweredIncorrectly();
        });
    });
});
