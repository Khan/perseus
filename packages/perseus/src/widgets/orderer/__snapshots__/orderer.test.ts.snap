// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderer widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Without using a calculator, put the numbers in order  from least to greatest.
        </strong>
          
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="draggy-boxy-thing orderer height-normal layout-horizontal above-scratchpad blank-background perseus-clearfix perseus-interactive"
          >
            <div
              class="bank perseus-clearfix"
            >
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card stack"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        1
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card stack"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        3
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card stack"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        2
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="perseus-clearfix draggable-box"
            >
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card drag-hint"
                />
              </div>
              <div />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`orderer widget should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Without using a calculator, put the numbers in order  from least to greatest.
        </strong>
          
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="draggy-boxy-thing orderer height-normal layout-horizontal above-scratchpad blank-background perseus-clearfix perseus-interactive"
          >
            <div
              class="bank perseus-clearfix"
            >
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card stack"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        1
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card stack"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        3
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card stack"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        2
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="perseus-clearfix draggable-box"
            >
              <div
                class="card-wrap perseus-interactive"
              >
                <div
                  class="card drag-hint"
                />
              </div>
              <div />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
